src/persist/src/gen.rs

```diff
@@ -0,0 +1,152 @@
+    ///
+    /// ```none
+    /// encoding = 6u8 v6_encoding
+    /// v6_encoding = md5_checksum proto_meta
+    /// md5_checksum = u8 u8 u8 u8 (little endian, md5 of proto_meta)
+    /// proto_meta = u8* (the protobuf serialization of ProtoMeta)
+    /// ```
+    pub const ENCODING_VERSION: u8 = 7;
+
+    /// The [Self::ENCODING_VERSION] of this previously encoded ProtoMeta.
+    ///
+    /// Returns an error if the input is malformed.
+    pub fn encoded_version(buf: &[u8]) -> Result<u8, Error> {
+        buf.get(0)
+            .copied()
+            .ok_or_else(|| Error::from("missing encoding version"))
+    }
+
+    // NB: This len is intentionally hardcoded (not derived from the md5 crate)
+    // so that a change to the crate can't break us. The compiler statically
+    // checks that they match because this const is used in the return type.
+    // (MD5 is not going to change, so this is all a moot point, but still
+    // better to be defensive.)
+    const CHECKSUM_LEN: usize = 16;
+    fn md5_checksum(buf: &[u8]) -> [u8; Self::CHECKSUM_LEN] {
+        let mut h = Md5::new();
+        h.update(&buf);
+        h.finalize().into()
+    }
+}
+
+impl persist_types::Codec for ProtoMeta {
+    fn codec_name() -> &'static str {
+        "protobuf+md5[ProtoMeta]"
+    }
+
+    fn size_hint(&self) -> usize {
+        // TODO: The encode step ends up internally calling compute_size a
+        // second time, which is unfortunate. Is it worth using
+        // write_to_with_cached_sizes in encode?
+        1 + Self::CHECKSUM_LEN + usize::cast_from(self.compute_size())
+    }
+
+    fn encode<E: for<'a> Extend<&'a u8>>(&self, buf: &mut E) {
+        let temp = self
+            .write_to_bytes()
+            .expect("no required fields means no initialization errors");
+        let checksum = Self::md5_checksum(&temp);
+        buf.extend(&[Self::ENCODING_VERSION]);
+        buf.extend(&checksum);
+        buf.extend(temp.iter());
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        let mut buf = buf;
+        let mut version = [0u8; 1];
+        let mut checksum = [0u8; Self::CHECKSUM_LEN];
+        match buf.read_exact(&mut version) {
+            Ok(_) if version == [Self::ENCODING_VERSION] => {}
+            Ok(_) => return Err(format!("unsupported version: {}", version[0])),
+            Err(_) => return Err("missing version".into()),
+        }
+        match buf.read_exact(&mut checksum) {
+            Ok(_) => {}
+            Err(_) => return Err("missing/incomplete checksum".into()),
+        }
+        if checksum != Self::md5_checksum(&buf) {
+            return Err("checksum mismatch".into());
+        }
+
+        // TODO: Use parse_from_carllerche_bytes to save allocs?
+        Self::parse_from_bytes(buf).map_err(|err| err.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use persist_types::Codec;
+
+    use super::*;
+
+    #[test]
+    fn checksum() {
```

Comment:
while I was reviewing this I was having a hard time convincing myself whether its sufficient to just test mutating the encoded data (and keep the persisted checksum the same) vs mutating the checksum, and keeping the encoded data are the same.

I think those two cases are symmetric and check the exact same code (read the bytes, separate out old checksum and data, compute new checksum, and compare with old) but wasn't 100% sure and wanted to flag that maybe adding a second test is worth it?

Reply:
good idea! I'll add that